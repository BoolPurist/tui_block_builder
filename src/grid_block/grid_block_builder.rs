use super::GridBlock;
use ndarray::Array2;

/// Builder for constructing a immutable `GridBlock`
/// # Example
/// ```
/// use block_builder::grid_block::BlockGridBuilder;
/// // Configure the grid to build
///  let block_grid =
///      // Every not set element has white space as value
///      BlockGridBuilder::with_default(' ')
///      // Every block contains be 2*2 elements
///      .block_size(2)
///      // Grid is 3 blocks wide
///      .blocks_in_x(3)
///      // Grid is 3 blocks high.
///      .blocks_in_y(3)
///      // elements of top left corner block have * as value
///      .set_block_sector(0, 0, '*')
///      // elements of bottom right corner block have * as value
///      .set_block_sector(2, 2, '*')
///      // Build grid according to current configuration
///      .build();
///
/// // Access every element by rows from top left to bottom right.
/// let mut rows: Vec<Vec<char>> = Vec::new();
/// for row in block_grid.iter() {
///     let mut to_push: Vec<char> = Vec::new();
///     for &column in row {
///         // do something with element
///         to_push.push(column);
///     }
///     // Do something the elements in a row
///     rows.push(to_push);
/// }
///
/// assert_eq!(
///     rows,
///     vec![
///         ['*', '*', ' ', ' ', ' ', ' ',],
///         ['*', '*', ' ', ' ', ' ', ' ',],
///         [' ', ' ', ' ', ' ',' ', ' ',],
///         [' ', ' ', ' ', ' ', ' ', ' ',],
///         [' ', ' ', ' ', ' ', '*', '*',],
///         [' ', ' ', ' ', ' ', '*', '*',],
///     ]
/// )
/// ```
pub struct BlockGridBuilder<T> {
    block_size: usize,
    blocks_in_x: usize,
    blocks_in_y: usize,
    default_value: T,
    setting_blocks: Vec<(usize, usize, T)>,
}

impl<T> Default for BlockGridBuilder<T>
where
    T: Default + Clone,
{
    fn default() -> Self {
        Self::init(Default::default())
    }
}

impl<T> BlockGridBuilder<T>
where
    T: Clone,
{
    /// Every not set element with have the value `default_value`
    pub fn with_default(default_value: T) -> Self {
        Self::init(default_value)
    }
    /// How big a block in the grid is.
    /// Grid will be [`block_size`] * [`blocks_in_y`] * [`blocks_in_x`] elements large.
    ///
    /// # Panic
    ///
    /// If `new_block_size` is zero
    pub fn block_size(&mut self, new_block_size: usize) -> &mut Self {
        if new_block_size < 1 {
            panic!("{} must be not be zero", stringify!(block_size));
        }

        self.block_size = new_block_size;

        self
    }

    pub fn reset_sectors(&mut self) -> &mut Self {
        self.setting_blocks.clear();

        self
    }

    /// All values in the block will be `value`.
    /// Example: given (1, 1) with * as value => value at (2, 2), (3, 2), (2, 3) (3, 3) will be * in
    /// constructed grid with `block_size` of 2.
    pub fn set_bulk_sectors(&mut self, value: T, sectors: &[(usize, usize)]) -> &mut Self {
        for (x, y) in sectors {
            self.setting_blocks.push((*x, *y, value.clone()));
        }

        self
    }

    /// # Panic
    ///
    /// If [`blocks_in_x`] and [`blocks_in_y`] are both zero after calling this method.
    pub fn blocks_in_x(&mut self, new_blocks_in_x: usize) -> &mut Self {
        self.validate_x_y();

        self.blocks_in_x = new_blocks_in_x;

        self
    }

    /// # Panic
    ///
    /// If [`blocks_in_x`] and [`blocks_in_y`] are both zero after calling this method.
    pub fn blocks_in_y(&mut self, new_blocks_in_y: usize) -> &mut Self {
        self.validate_x_y();

        self.blocks_in_y = new_blocks_in_y;

        self
    }

    pub fn set_block_sector(&mut self, x: usize, y: usize, value: T) -> &mut Self {
        self.setting_blocks.push((x, y, value));
        self
    }

    pub fn build(&self) -> GridBlock<T> {
        let block_size = self.block_size;
        let width = block_size * self.blocks_in_x;
        let height = block_size * self.blocks_in_y;
        let mut build: Array2<T> = Array2::from_elem((height, width), self.default_value.clone());

        for (x, y, to_insert) in self.setting_blocks.iter() {
            let to_add = block_size - 1;

            let (scaled_x, scaled_y) = (*x * block_size, *y * block_size);

            let max_x = scaled_x + to_add;
            let max_y = scaled_y + to_add;

            for next_y in scaled_y..=max_y {
                for next_x in scaled_x..=max_x {
                    let insert_into = build
                        .get_mut((next_y, next_x))
                        .expect("Unexpected: during insertion of value out of bound indexing");
                    *insert_into = to_insert.clone();
                }
            }
        }

        GridBlock::new(build, width, height)
    }

    fn init(value: T) -> Self {
        Self {
            block_size: 1,
            blocks_in_x: 1,
            blocks_in_y: 1,
            setting_blocks: Default::default(),
            default_value: value,
        }
    }

    fn validate_x_y(&self) {
        if self.blocks_in_x < 1 && self.blocks_in_y < 1 {
            panic!(
                "{} and {} must be not be zero",
                stringify!(blocks_in_x),
                stringify!(blocks_in_y)
            );
        }
    }
}

#[cfg(test)]
mod testing {
    use super::*;
    #[test]
    fn should_set_sectors() {
        let block = BlockGridBuilder::with_default(' ')
            .blocks_in_y(3)
            .blocks_in_x(4)
            .block_size(2)
            .set_block_sector(0, 0, '*')
            .set_block_sector(1, 0, '*')
            .set_block_sector(2, 0, '*')
            .set_block_sector(2, 1, '*')
            .set_block_sector(2, 2, '*')
            .set_block_sector(3, 0, '*')
            .set_block_sector(0, 1, '*')
            .build();

        insta::assert_display_snapshot!(block);
    }

    #[test]
    #[rustfmt::skip]
    fn doc_test() {
        
   }

    #[test]
    fn should_should_produce_nine() {
        let block = create_base_nine().build();

        insta::assert_display_snapshot!(block);
    }
    #[test]
    fn should_should_produce_nine_x3() {
        let block = create_base_nine().block_size(3).build();

        insta::assert_display_snapshot!(block);
    }

    fn create_base_nine() -> BlockGridBuilder<char> {
        let mut block = BlockGridBuilder::with_default(' ');
        block
            .blocks_in_y(5)
            .blocks_in_x(3)
            .block_size(1)
            // ***
            // * *
            // ***
            //   *
            // ***
            //
            // ***
            .set_block_sector(0, 0, '*')
            .set_block_sector(1, 0, '*')
            .set_block_sector(2, 0, '*')
            // * *
            .set_block_sector(0, 1, '*')
            .set_block_sector(2, 1, '*')
            // ***
            .set_block_sector(0, 2, '*')
            .set_block_sector(1, 2, '*')
            .set_block_sector(2, 2, '*')
            //   *
            .set_block_sector(2, 3, '*')
            // ***
            .set_block_sector(0, 4, '*')
            .set_block_sector(1, 4, '*')
            .set_block_sector(2, 4, '*');
        block
    }
}
