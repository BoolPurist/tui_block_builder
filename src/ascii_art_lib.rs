//! Provides helper functions to build block like numbers, : and empty white space seperator

use crate::grid_block::BlockGridBuilder;

const TOP_LEFT: (usize, usize) = (0, 0);
const TOP_CENTER: (usize, usize) = (1, 0);
const TOP_RIGHT: (usize, usize) = (2, 0);

const UPPER_LEFT: (usize, usize) = (0, 1);
const UPPER_RIGHT: (usize, usize) = (2, 1);

const MIDDLE_LEFT: (usize, usize) = (0, 2);
const MIDDLE_CENTER: (usize, usize) = (1, 2);
const MIDDLE_RIGHT: (usize, usize) = (2, 2);

const LOWER_LEFT: (usize, usize) = (0, 3);
const LOWER_RIGHT: (usize, usize) = (2, 3);

const BOTTOM_LEFT: (usize, usize) = (0, 4);
const BOTTOM_CENTER: (usize, usize) = (1, 4);
const BOTTOM_RIGHT: (usize, usize) = (2, 4);

pub fn build_1<T>(default_v: T, taken_v: T) -> BlockGridBuilder<T>
where
    T: Clone,
{
    let mut builder = build_thin_number_base(default_v);
    builder.reset_sectors().set_bulk_sectors(
        taken_v,
        &[
            TOP_RIGHT,
            UPPER_RIGHT,
            MIDDLE_RIGHT,
            LOWER_RIGHT,
            BOTTOM_RIGHT,
        ],
    );
    builder
}

#[rustfmt::skip]
pub fn build_2<T>(default_v: T, taken_v: T) -> BlockGridBuilder<T>
where
    T: Clone,
{
    let mut builder = build_thin_number_base(default_v);
    builder.set_bulk_sectors(
        taken_v,
        &[
            TOP_LEFT,    TOP_CENTER,    TOP_RIGHT,
                                        UPPER_RIGHT,
            MIDDLE_LEFT, MIDDLE_CENTER, MIDDLE_RIGHT,
            LOWER_LEFT,
            BOTTOM_LEFT, BOTTOM_CENTER, BOTTOM_RIGHT,
        ],
    );
    builder
}

#[rustfmt::skip]
pub fn build_3<T>(default_v: T, taken_v: T) -> BlockGridBuilder<T>
where
    T: Clone,
{
    let mut builder = build_thin_number_base(default_v);
    builder.set_bulk_sectors(
        taken_v,
        &[
            TOP_LEFT,    TOP_CENTER,    TOP_RIGHT,
                                        UPPER_RIGHT,
            MIDDLE_LEFT, MIDDLE_CENTER, MIDDLE_RIGHT,
                                        LOWER_RIGHT,
            BOTTOM_LEFT, BOTTOM_CENTER, BOTTOM_RIGHT,
        ],
    );
    builder
}

#[rustfmt::skip]
pub fn build_4<T>(default_v: T, taken_v: T) -> BlockGridBuilder<T>
where
    T: Clone,
{
    let mut builder = build_thin_number_base(default_v);
    builder.set_bulk_sectors(
        taken_v,
        &[
            TOP_LEFT,                   TOP_RIGHT,
            UPPER_LEFT,                 UPPER_RIGHT,
            MIDDLE_LEFT, MIDDLE_CENTER, MIDDLE_RIGHT,
                                        LOWER_RIGHT,
                                        BOTTOM_RIGHT,
        ],
    );
    builder
}
#[rustfmt::skip]
pub fn build_5<T>(default_v: T, taken_v: T) -> BlockGridBuilder<T>
where
    T: Clone,
{
    let mut builder = build_thin_number_base(default_v);
    builder.set_bulk_sectors(
        taken_v,
        &[
            TOP_LEFT,    TOP_CENTER,    TOP_RIGHT,
            UPPER_LEFT,                 
            MIDDLE_LEFT, MIDDLE_CENTER, MIDDLE_RIGHT,
                                        LOWER_RIGHT,
            BOTTOM_LEFT, BOTTOM_CENTER, BOTTOM_RIGHT,
        ],
    );
    builder
}
#[rustfmt::skip]
pub fn build_6<T>(default_v: T, taken_v: T) -> BlockGridBuilder<T>
where
    T: Clone,
{
    let mut builder = build_thin_number_base(default_v);
    builder.set_bulk_sectors(
        taken_v,
        &[
            TOP_LEFT,    TOP_CENTER,    TOP_RIGHT,
            UPPER_LEFT,                 
            MIDDLE_LEFT, MIDDLE_CENTER, MIDDLE_RIGHT,
            LOWER_LEFT,                 LOWER_RIGHT,
            BOTTOM_LEFT, BOTTOM_CENTER, BOTTOM_RIGHT,
        ],
    );
    builder
}
#[rustfmt::skip]
pub fn build_7<T>(default_v: T, taken_v: T) -> BlockGridBuilder<T>
where
    T: Clone,
{
    let mut builder = build_thin_number_base(default_v);
    builder.set_bulk_sectors(
        taken_v,
        &[
            TOP_LEFT,    TOP_CENTER,    TOP_RIGHT,
                                        UPPER_RIGHT, 
                                        MIDDLE_RIGHT,
                                        LOWER_RIGHT,
                                        BOTTOM_RIGHT,
        ],
    );
    builder
}
#[rustfmt::skip]
pub fn build_8<T>(default_v: T, taken_v: T) -> BlockGridBuilder<T>
where
    T: Clone,
{
    let mut builder = build_thin_number_base(default_v);
    builder.set_bulk_sectors(
        taken_v,
        &[
            TOP_LEFT,    TOP_CENTER,    TOP_RIGHT,
            UPPER_LEFT,                 UPPER_RIGHT, 
            MIDDLE_LEFT, MIDDLE_CENTER, MIDDLE_RIGHT,
            LOWER_LEFT,                 LOWER_RIGHT,
            BOTTOM_LEFT, BOTTOM_CENTER, BOTTOM_RIGHT,
        ],
    );
    builder
}
#[rustfmt::skip]
pub fn build_9<T>(default_v: T, taken_v: T) -> BlockGridBuilder<T>
where
    T: Clone,
{
    let mut builder = build_thin_number_base(default_v);
    builder.set_bulk_sectors(
        taken_v,
        &[
            TOP_LEFT,    TOP_CENTER,    TOP_RIGHT,
            UPPER_LEFT,                 UPPER_RIGHT, 
            MIDDLE_LEFT, MIDDLE_CENTER, MIDDLE_RIGHT,
                                        LOWER_RIGHT,
            BOTTOM_LEFT, BOTTOM_CENTER, BOTTOM_RIGHT,
        ],
    );
    builder
}
#[rustfmt::skip]
pub fn build_0<T>(default_v: T, taken_v: T) -> BlockGridBuilder<T>
where
    T: Clone,
{
    let mut builder = build_thin_number_base(default_v);
    builder.set_bulk_sectors(
        taken_v,
        &[
            TOP_LEFT,    TOP_CENTER,    TOP_RIGHT,
            UPPER_LEFT,                 UPPER_RIGHT, 
            MIDDLE_LEFT,                MIDDLE_RIGHT,
            LOWER_LEFT,                 LOWER_RIGHT,
            BOTTOM_LEFT, BOTTOM_CENTER, BOTTOM_RIGHT,
        ],
    );
    builder
}
pub fn build_double_point<T>(default_v: T, taken_v: T) -> BlockGridBuilder<T>
where
    T: Clone,
{
    let mut builder = build_thin_number_base(default_v);
    builder.set_bulk_sectors(taken_v, &[(1, 1), (1, 3)]);
    builder
}
pub fn build_space<T>(default_v: T) -> BlockGridBuilder<T>
where
    T: Clone,
{
    let mut builder = BlockGridBuilder::with_default(default_v);
    builder.block_size(1).blocks_in_x(1).blocks_in_y(3);
    builder
}

pub fn build_thin_number_base<T>(default_v: T) -> BlockGridBuilder<T>
where
    T: Clone,
{
    let mut builder = BlockGridBuilder::with_default(default_v);
    builder.block_size(1).blocks_in_x(3).blocks_in_y(5);
    builder
}

#[cfg(test)]
mod testing {
    use super::*;
    #[test]
    fn should_build_1() {
        let one = build_1(' ', '*').build();
        insta::assert_display_snapshot!(one);
    }
    #[test]
    fn should_build_2() {
        let two = build_2(' ', '*').build();
        insta::assert_display_snapshot!(two);
    }
    #[test]
    fn should_build_3() {
        let two = build_3(' ', '*').build();
        insta::assert_display_snapshot!(two);
    }
    #[test]
    fn should_build_4() {
        let four = build_4(' ', '*').build();
        insta::assert_display_snapshot!(four);
    }
    #[test]
    fn should_build_5() {
        let five = build_5(' ', '*').build();
        insta::assert_display_snapshot!(five);
    }
    #[test]
    fn should_build_6() {
        let six = build_6(' ', '*').build();
        insta::assert_display_snapshot!(six);
    }
    #[test]
    fn should_build_7() {
        let seven = build_7(' ', '*').build();
        insta::assert_display_snapshot!(seven);
    }
    #[test]
    fn should_build_8() {
        let eight = build_8(' ', '*').build();
        insta::assert_display_snapshot!(eight);
    }
    #[test]
    fn should_build_9() {
        let nine = build_9(' ', '*').build();
        insta::assert_display_snapshot!(nine);
    }
    #[test]
    fn should_build_0() {
        let zero = build_0(' ', '*').build();
        insta::assert_display_snapshot!(zero);
    }
    #[test]
    fn should_build_double_point() {
        let double_point = build_double_point(' ', '*').build();
        insta::assert_display_snapshot!(double_point);
    }
    #[test]
    fn should_space() {
        let space = build_space(' ').build();
        insta::assert_display_snapshot!(space);
    }
}
