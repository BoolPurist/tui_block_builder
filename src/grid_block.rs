//! Grid system to composite blocks

use core::fmt::{Debug, Display};
use ndarray::Array2;

mod grid_block_builder;
pub use grid_block_builder::BlockGridBuilder;

/// Grid which is made up in blocks . Every block has the same size. A block has size * size
/// elements.
/// Every grid has a number of blocks in x and y direction.
/// The whole API is exposed for immutable access only.
#[derive(Debug)]
pub struct GridBlock<T> {
    grid: Array2<T>,
    width: usize,
    height: usize,
}

impl<T> GridBlock<T> {
    fn new(grid: Array2<T>, width: usize, height: usize) -> Self {
        Self {
            grid,
            width,
            height,
        }
    }
}

impl<T> GridBlock<T> {
    /// Number of elements in x direction
    pub fn width(&self) -> usize {
        self.width
    }

    /// Number of elements in y direction
    pub fn height(&self) -> usize {
        self.height
    }

    pub fn get(&self, x: usize, y: usize) -> Option<&T> {
        self.grid.get((y, x))
    }

    /// Get all elements along a y axis
    pub fn get_row_at(&self, y: usize) -> Option<&[T]> {
        self.grid.row(y).to_slice()
    }

    /// Yields all elements of blocks given in param `grid` along on given y axis `y`
    /// This operation can be used to traverse by rows as if the given block are one.
    pub fn iter_along_x_axis(grid: &[GridBlock<T>], y: usize) -> impl Iterator<Item = &T> {
        grid.iter()
            .filter_map(move |next_grid| next_grid.get_row_at(y))
            .flat_map(|row| row.iter())
    }

    pub fn iter(&self) -> impl Iterator<Item = &[T]>
    where
        T: Clone,
    {
        (0..self.height()).map(|index| self.get_row_at(index).unwrap())
    }
    /// Yields elements of given `grid` for traversing from top left to bottom right corner by
    /// rows. It stops at row `max_len`.
    pub fn iter_top_left_bottom_right(
        grid: &[GridBlock<T>],
        max_len: usize,
    ) -> impl Iterator<Item = Vec<&T>> {
        (0..max_len).map(|index| Self::iter_along_x_axis(grid, index).collect())
    }
}

impl<T> Display for GridBlock<T>
where
    T: Display,
{
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.grid)
    }
}

#[cfg(test)]
mod testing {
    use super::*;
    use crate::ascii_art_lib;
    #[test]
    fn should_get_row_at() {
        const UNEXPECTED_MSG: &str = "No zeroth row for given block in test";
        let block = BlockGridBuilder::with_default(' ')
            .block_size(2)
            .blocks_in_x(3)
            .blocks_in_y(3)
            .set_block_sector(0, 0, '*')
            .set_block_sector(2, 0, '^')
            .set_block_sector(0, 1, 'x')
            .build();

        let row = block.get_row_at(0).expect(UNEXPECTED_MSG);

        let third_row = block.get_row_at(2).expect(UNEXPECTED_MSG);
        insta::assert_yaml_snapshot!("1 and 5 row", &[row, third_row]);
    }

    #[test]
    fn should_traverse_from_top_left_to_bottom_right() {
        let one = ascii_art_lib::build_1(" ", "*").build();
        let double_point = ascii_art_lib::build_double_point(" ", "x").build();
        let four = ascii_art_lib::build_4(" ", "*").build();

        let actual: Vec<String> =
            GridBlock::iter_top_left_bottom_right(&vec![one, double_point, four], 5)
                .map(|line| {
                    line.into_iter()
                        .map(|symbol| symbol.to_string())
                        .collect::<Vec<String>>()
                        .join("")
                })
                .collect();

        insta::assert_yaml_snapshot!(actual);
    }
}
