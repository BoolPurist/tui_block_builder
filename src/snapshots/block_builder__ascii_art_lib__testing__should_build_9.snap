---
source: src/ascii_art_lib.rs
expression: nine
---
[[*, *, *],
 [*,  , *],
 [*, *, *],
 [ ,  , *],
 [*, *, *]]
